lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php

```diff
@@ -0,0 +1,500 @@
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exec($statement)
+    {
+        $stmt = $this->prepare($statement);
+        $stmt->execute();
+
+        return $stmt->rowCount();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastInsertId($name = null)
+    {
+        if ($name === null) {
+            return false;
+        }
+
+        $sql = 'SELECT GEN_ID(' . $name . ', 0) LAST_VAL FROM RDB$DATABASE';
+        $stmt = $this->query($sql);
+        $result = $stmt->fetchColumn(0);
+
+        return $result;
+    }
+
+    /**
+     * Returns the current execution mode.
+     *
+     * @return integer
+     */
+    public function getExecuteMode()
+    {
+        return $this->executeMode;
+    }
+
+    /**
+     * Generates an SET TRANSACTION statement used to start an transaction
+     * 
+     * @param type $isolationLevel
+     * @param type $timeout
+     * @return string
+     */
+    public function getStartTransactionSql($isolationLevel, $timeout = 5)
+    {
+        $result = "";
+        switch ($isolationLevel) {
+            case \Doctrine\DBAL\Connection::TRANSACTION_READ_UNCOMMITTED: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL READ COMMITTED RECORD_VERSION';
+                    break;
+                }
+            case \Doctrine\DBAL\Connection::TRANSACTION_READ_COMMITTED: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL READ COMMITTED RECORD_VERSION';
+                    break;
+                }
+            case \Doctrine\DBAL\Connection::TRANSACTION_REPEATABLE_READ: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL SNAPSHOT ';
+                    break;
+                }
+            case \Doctrine\DBAL\Connection::TRANSACTION_SERIALIZABLE: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL SNAPSHOT TABLE STABILITY';
+                    break;
+                }
+        }
+
+        if (($this->attrDcTransWait > 0)) {
+            $result .= ' WAIT LOCK TIMEOUT ' . $this->attrDcTransWait;
+        } elseif  (($this->attrDcTransWait === -1)) {
+            $result .= ' WAIT';
+        } else {
+            $result .= ' NO WAIT';
+        }
+
+        return $result;
+    }
+
+    /**
+     * Starts a new transaction and returns the transaction handle
+     * 
+     * @param resource $commitDefaultTransaction
+     */
+    protected function internalBeginTransaction($commitDefaultTransaction = true)
+    {
+        if ($commitDefaultTransaction) {
+            @ibase_commit($this->ibaseConnectionRc);
+        }
+        $result = @ibase_query($this->ibaseConnectionRc, $this->getStartTransactionSql($this->attrDcTransIsolationLevel));
+        if (!is_resource($result)) {
+            $this->checkLastApiCall();
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     * 
+     */
+    public function beginTransaction()
```

Comment:
Can you explain all this transaction code for?
At first glance it seems to support nested transactions, but  i fail see how to reach a transaction depth > 1.